File: aoc-py/solutions/day25.py
"""
Day 25: Snowverload

https://adventofcode.com/2023/day/25
"""
__all__ = ('Day25',)

from typing import ClassVar

import networkx as nx

from ..solution import Solution

class Day25(Solution):
    NAME: ClassVar[str] = 'Snowverload'

    def part_one(self, inp: str) -> int:
        graph = nx.Graph()

        # generate graph
        for line in inp.splitlines():
            left, right = line.split(':')
            for node in right.strip().split():
                graph.add_edge(left, node)
                graph.add_edge(node, left)

        # remove the edges that are of the cut
        graph.remove_edges_from(
            nx.minimum_edge_cut(graph)
        )

        # get 2 partitioned groups after the cut (`a` and `b`)
        a, b = nx.connected_components(graph)
        # multiply sizes of 2 groups together
        return len(a) * len(b)

    def part_two(self, _: str) -> None:
        """No part 2 for day 25!

        Merry Christmas!
        """

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))

        assert p1 == 554064

File: src/bin/day25.rs
//! Day 25: Snowverload
//!
//! <https://adventofcode.com/2023/day/25>
use std::fmt::Display;
use rustworkx_core::{
    petgraph::graphmap::GraphMap,
    connectivity::stoer_wagner_min_cut as min_cut,
};
use aoc_2023::Solution;

pub struct Day25;

impl Day25 {
    pub fn part_one<T: Display>(&self, inp: T) -> usize {
        let mut graph = GraphMap::new();
        let inp = inp
            .to_string();

        for line in inp.lines() {
            let (left, right) = line
                .split_once(':')
                .unwrap();
            for node in right
                .trim()
                .split_whitespace()
            {
                graph.add_edge(left, node, 1);
                graph.add_edge(node, left, 1);
            }
        }

        if let Ok(Some((_, partition_1))) = min_cut(
            &graph,
            |_| Ok::<usize, usize>(1))
        {
            let size_1 = partition_1.len();
            let size_2 = graph.node_count() - size_1;

            size_1 * size_2
        } else {
            panic!("Could not perform a minimum cut on the graph");
        }
    }

    pub fn part_two<T: Display>(&self, _inp: T) -> usize {
        unimplemented!()
    }
}

impl Solution for Day25 {
    const NAME: &'static str = "Snowverload";

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);

        println!("Part 1: {p1}");

        assert_eq!(p1, 554064);
    }
}

fn main() {
    aoc_2023::run_day(25, &Day25);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}